pkg/beholder/client.go

```diff
@@ -0,0 +1,312 @@
+package beholder
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	otellog "go.opentelemetry.io/otel/log"
+	otelglobal "go.opentelemetry.io/otel/log/global"
+	otelmetric "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/propagation"
+	sdklog "go.opentelemetry.io/otel/sdk/log"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	sdkresource "go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/sdk/trace"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+type Emitter interface {
+	// Sends message with bytes and attributes to OTel Collector
+	Emit(ctx context.Context, body []byte, attrs map[string]any) error
+	// Sends message to OTel Collector
+	EmitMessage(ctx context.Context, m Message) error
+}
+type Client interface {
+	Logger() otellog.Logger
+	Tracer() oteltrace.Tracer
+	Meter() otelmetric.Meter
+	Emitter() Emitter
+	Close() error
+}
+
+var _ Client = (*beholderClient)(nil)
+
+type messageEmitter struct {
+	exporter      sdklog.Exporter
+	messageLogger otellog.Logger
+	retryCount    uint
+	retryDelay    time.Duration
+}
+
+type beholderClient struct {
+	config Config
+	// Logger
+	logger otellog.Logger
+	// Tracer
+	tracer oteltrace.Tracer
+	// Meter
+	meter otelmetric.Meter
+	// MessageEmitter
+	messageEmitter Emitter
+	// Graceful shutdown for tracer, meter, logger providers
+	closeFunc func() error
+}
+
+func NewClient(
+	config Config,
+	logger otellog.Logger,
+	tracer oteltrace.Tracer,
+	meter otelmetric.Meter,
+	messageEmitter Emitter,
+	onClose func() error,
+) Client {
+	return &beholderClient{
+		config:         config,
+		logger:         logger,
+		tracer:         tracer,
+		meter:          meter,
+		messageEmitter: messageEmitter,
+		closeFunc:      onClose,
+	}
+}
+
+// NewOtelClient creates a new BeholderClient with OTel exporter
+func NewOtelClient(cfg Config, errorHandler errorHandlerFunc) (Client, error) {
+	factory := func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error) {
+		return otlploggrpc.New(ctx, options...)
+	}
+	return newOtelClient(cfg, errorHandler, factory)
+}
+
+// Used for testing to override the default exporter
+type otlploggrpcFactory func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error)
+
+func newOtelClient(cfg Config, errorHandler errorHandlerFunc, otlploggrpcNew otlploggrpcFactory) (Client, error) {
+	ctx := context.Background()
+	baseResource, err := newOtelResource()
+	if err != nil {
+		return nil, err
+	}
+	sharedLogExporter, err := otlploggrpcNew(
+		ctx,
+		otlploggrpc.WithInsecure(),
```

Comment:
should this option be configurable? 

(also applies to the `Meter` and `Tracer` configs later on in this file)

Reply:
Addressed [here](https://github.com/smartcontractkit/chainlink-common/pull/681/commits/0bdc6d2b4de8da9acc5ce6c3950060f26a09d061)